src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreePostingsWriter.java

```diff
@@ -0,0 +1,215 @@
+ * This allows us to skip reading the leaves, instead just order leaf blocks by their offset in the index file,
+ * and correlate them with buffered posting lists.
+ */
+@NotThreadSafe
+public class BlockBalancedTreePostingsWriter implements BlockBalancedTreeWalker.TraversalCallback
+{
+    private static final Logger logger = LoggerFactory.getLogger(BlockBalancedTreePostingsWriter.class);
+
+    /**
+     * Minimum number of reachable leaves for a given node to be eligible for an auxiliary posting list.
+     */
+    public static final int MINIMUM_POSTINGS_LEAVES = CassandraRelevantProperties.SAI_MINIMUM_POSTINGS_LEAVES.getInt();
+    /**
+     * Skip, or the sampling interval, for selecting a balanced tree level that is eligible for an auxiliary posting list.
+     * Sampling starts from 0, but the balanced tree root node is at level 1. For skip = 4, eligible levels are 4, 8, 12, etc. (no
+     * level 0, because there is no node at level 0).
+     */
+    public static final int POSTINGS_SKIP = CassandraRelevantProperties.SAI_POSTINGS_SKIP.getInt();
+
+    private final List<PackedLongValues> postings;
+    private final TreeMap<Long, Integer> leafOffsetToNodeID = new TreeMap<>(Long::compareTo);
+    private final Multimap<Integer, Integer> nodeToChildLeaves = HashMultimap.create();
+
+    private final IndexContext indexContext;
+    int numNonLeafPostings = 0;
+    int numLeafPostings = 0;
+
+    BlockBalancedTreePostingsWriter(List<PackedLongValues> postings, IndexContext indexContext)
+    {
+        this.postings = postings;
+        this.indexContext = indexContext;
+    }
+
+    /**
+     * Called when a leaf node is hit as we traverse the packed index.
+     *
+     * @param leafNodeID the current leaf node ID in the packed inded
+     * @param leafBlockFP the file pointer to the on-disk leaf block
+     * @param pathToRoot the path to the root leaf above this leaf. Contains all the intermediate leaf node IDs.
+     */
+    @Override
+    public void onLeaf(int leafNodeID, long leafBlockFP, IntArrayList pathToRoot)
+    {
+        checkArgument(!pathToRoot.containsInt(leafNodeID));
+        checkArgument(pathToRoot.isEmpty() || leafNodeID > pathToRoot.get(pathToRoot.size() - 1));
+
+        leafOffsetToNodeID.put(leafBlockFP, leafNodeID);
+        for (int i = 0; i < pathToRoot.size(); i++)
+        {
+            int level = i + 1;
+            if (isLevelEligibleForPostingList(level))
+            {
+                int nodeID = pathToRoot.get(i);
+                nodeToChildLeaves.put(nodeID, leafNodeID);
+            }
+        }
+    }
+
+    /**
+     * Writes a merged posting list for each leaf in the tree. This postings list consists of the postings
+     * associated with the leaf along with the postings associated with any leaves underneath it.
+     *
+     * After writing out the postings it writes a map of leaf node IDs to postings file pointer for all
```

Comment:
nit: The map contains both internal node IDs and leaf node IDs?

Alternate wording...

```
* After writing out the postings, it writes a map of node ID -> postings file pointer for all
* nodes with an attached postings list. It then returns the file pointer to this map.
 ```